### PR TITLE
Support statistics in pg_stat* views for auxiliary relations of AO table

### PR DESCRIPTION
**Pull request by maksm90**
The auxiliary system relations of AO tables such as aosegment, block
directory and visibility map have different from regular heap tables
relkind values. As a consequence backend processes don't gather
statistics for these tables on tuple and buffer accesses/modifications.
And therefore there isn't any facility to see this statistics from
pg_stat* views.

The current fix derestricts this feature. Also the light fix that
segregates HOT-updates from regular "cold" ones is incorporated when
after updating of heap tuple the backend process sums up corresponing
counters into local cache.

### CODE REVIEW

File: src/test/regress/expected/aux_ao_rels_stat.out [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+-- Wait for stats collector to update accumulated statistics
+DO $$
+DECLARE
+    start_time timestamptz := clock_timestamp();
+    updated bool;
+BEGIN
+    FOR i IN 1 .. 300 LOOP
+        SELECT (sum(seq_scan) > 0) INTO updated
+        FROM gp_dist_random('pg_stat_user_tables') pgsut
+        JOIN pg_appendonly pgao ON pgsut.relid = pgao.segrelid
```
Comment (darthunix) [inline src/test/regress/expected/aux_ao_rels_stat.out:32]: Do we have guarantees that `segrelid` is equal among all segments? If not, we should also wrap `pg_appendonly` with `gpdist_random`.

**Review by darthunix — CHANGES_REQUESTED** [review]
1. `aux_ao_rels_stat` test fails on my laptop. I've found out that `pg_stat_all_tables` doesn't contain any information about auxiliary `pg_aoseg.pg_aocsseg_XXX` relation.
2. I have also found out that we can't run analyze on auxiliary tables as their relkind differs from `RELKIND_RELATION`. An error raises in [analyze.c](https://github.com/arenadata/gpdb/blob/ef79c2a0381cd0906a0544a971e4abcf988c82b8/src/backend/commands/analyze.c#L380).

File: src/test/regress/expected/aux_ao_rels_stat.out (hunk 1 above)
Comment (maksm90) [reply src/test/regress/expected/aux_ao_rels_stat.out:32]: AFAICS relation oids are the same for the same tables on master and segments

**Comment by maksm90** [comment]
@darthunix thanks for the review.

> aux_ao_rels_stat test fails on my laptop. I've found out that pg_stat_all_tables doesn't contain any information about auxiliary pg_aoseg.pg_aocsseg_XXX relation.

Hmm, could you share your gotten output after running test? Do you set up `track_counts` GUC?

> I have also found out that we can't run analyze on auxiliary tables as their relkind differs from RELKIND_RELATION. An error raises in analyze.c.

By now analyze is not applicable to auxiliary tables. But it doesn't have any relation to this task because the analyze forms planner statistics - not monitoring one that is the case of current task.

**Comment by maksm90** [comment]
> aux_ao_rels_stat test fails on my laptop. I've found out that pg_stat_all_tables doesn't contain any information about auxiliary pg_aoseg.pg_aocsseg_XXX relation.

The reason of this behavior also could be that you need to set up data directory from scratch because of patch contains fixes to the system views that are applied under `intidb` execution.
I suppose this factor is major for your case.

**Comment by darthunix** [comment]
@maksm90 thanks, you are right - I had to recreate demo cluster. Now `aux_ao_rels_stat` pass.

**Comment by darthunix** [comment]
I have mentioned about `analyze` just because was not sure is it correct behaviour or not. Surely it is not interconnected with `pg_stats`.

**Review by darthunix — APPROVED** [review]

**Review by leskin-in — APPROVED** [review]

**Comment by maksm90** [comment]
> I have mentioned about analyze just because was not sure is it correct behavior or not

GPDB 6X disables `analyze` for auxiliary relations. This could be confusing as `n_mod_since_analyze` counter of `n_mod_since_analyze` table isn't ever flushed to zero. But the corresponding statistics in `pg_statistics` and `pg_class` system relations is not counted for auxiliary relations too. Fortunately vacuum statistics (such as `last_vacuum` timestamp and `vacuum_count`) is good for these tables

**Review by Stolb27 — APPROVED** [review]

**Pull request merged**